src/cuda-poh-verify/poh_verify.cu

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+    }
+    return g_total_gpus > 0;
+}
+
+bool poh_init() {
+    cudaFree(0);
+    pthread_mutex_lock(&g_ctx_mutex);
+    bool success = poh_init_locked();
+    pthread_mutex_unlock(&g_ctx_mutex);
+    return success;
+}
+
+// This function locks the pthreads mutex for the gpu context it takes
+// for the computation. `poh_finish_verify_many` must be called with the
+// returned gpu_ctx to avoid poisoning the mutex. This structure allows
+// the host to complete calculations concurrently with the gpu calculations.
+gpu_ctx* poh_start_verify_many(const uint8_t* hashes,
+                                     const uint64_t* num_hashes_arr,
+                                     size_t num_elems,
+                                     uint8_t use_non_default_stream)
+{
+    LOG("Starting poh_verify_many: num_elems: %zu\n", num_elems);
+
+    int32_t cur_gpu, cur_queue;
+
+    pthread_mutex_lock(&g_ctx_mutex);
+    if (!poh_init_locked()) {
+        pthread_mutex_unlock(&g_ctx_mutex);
+        LOG("No GPUs, exiting...\n");
+        return NULL;
+    }
+    cur_gpu = g_cur_gpu;
+    g_cur_gpu++;
+    g_cur_gpu %= g_total_gpus;
+    cur_queue = g_cur_queue[cur_gpu];
+    g_cur_queue[cur_gpu]++;
+    g_cur_queue[cur_gpu] %= MAX_QUEUE_SIZE;
+    pthread_mutex_unlock(&g_ctx_mutex);
+
+    gpu_ctx* cur_ctx = &g_gpu_ctx[cur_gpu][cur_queue];
+    pthread_mutex_lock(&cur_ctx->mutex);
+
+    CUDA_CHK(cudaSetDevice(cur_gpu));
+
+    LOG("cur gpu: %d cur queue: %d\n", cur_gpu, cur_queue);
+
+    size_t hashes_size = num_elems * SHA256_BLOCK_SIZE * sizeof(uint8_t);
+    size_t num_hashes_size = num_elems * sizeof(uint64_t);
+
+    // Ensure there is enough memory allocated
+    if (cur_ctx->hashes == NULL || cur_ctx->num_elems_alloc < num_elems) {
+        CUDA_CHK(cudaFree(cur_ctx->hashes));
+        CUDA_CHK(cudaMalloc(&cur_ctx->hashes, hashes_size));
+        CUDA_CHK(cudaFree(cur_ctx->num_hashes_arr));
+        CUDA_CHK(cudaMalloc(&cur_ctx->num_hashes_arr, num_hashes_size));
+
+        cur_ctx->num_elems_alloc = num_elems;
+    }
+
+    cudaStream_t stream = 0;
+    if (0 != use_non_default_stream) {
+        stream = cur_ctx->stream;
+    }
+
+    CUDA_CHK(cudaMemcpyAsync(cur_ctx->hashes, hashes, hashes_size, cudaMemcpyHostToDevice, stream));
+    CUDA_CHK(cudaMemcpyAsync(cur_ctx->num_hashes_arr, num_hashes_arr, num_hashes_size, cudaMemcpyHostToDevice, stream));
+
+    int num_blocks = ROUND_UP_DIV(num_elems, NUM_THREADS_PER_BLOCK);
+
+    poh_verify_kernel<<<num_blocks, NUM_THREADS_PER_BLOCK, 0, stream>>>(cur_ctx->hashes, cur_ctx->num_hashes_arr, num_elems);
+    CUDA_CHK(cudaPeekAtLastError());
+
+    return cur_ctx;
+}
+
+void poh_finish_verify_many(uint8_t* hashes, size_t num_elems, gpu_ctx* cur_ctx, uint8_t use_non_default_stream) {
+    cudaStream_t stream = 0;
+    if (0 != use_non_default_stream) {
+        stream = cur_ctx->stream;
+    }
+
+    size_t hashes_size = num_elems * SHA256_BLOCK_SIZE * sizeof(uint8_t);
```

Comment:
Would be nice if this is not duplicated.

Reply:
Yeah, looking at this again, I might refactor it to only be one function, then have rust spawn a thread to call it while Rust does the tx hashing.